homie/src/lib.rs

```diff
@@ -282,13 +332,68 @@ impl HomieDevice {
     }
 
     /// Spawn a task to handle the EventLoop.
-    fn spawn(mut event_loop: EventLoop) -> JoinHandle<Result<(), Box<dyn Error + Send + Sync>>> {
-        task::spawn(async move {
+    fn spawn(
+        &self,
+        mut event_loop: EventLoop,
+        mut update_callback: Option<UpdateCallback>,
+    ) -> impl Future<Output = Result<(), Box<dyn Error + Send + Sync>>> {
+        let device_base = format!("{}/", self.publisher.device_base);
+        let (incoming_tx, incoming_rx) = async_channel::unbounded();
+
+        let mqtt_task = task::spawn(async move {
             loop {
                 let (incoming, outgoing) = event_loop.poll().await?;
                 log::trace!("Incoming = {:?}, Outgoing = {:?}", incoming, outgoing);
+
+                if let Some(incoming) = incoming {
+                    incoming_tx.send(incoming).await?;
+                }
             }
-        })
+        });
+
+        let publisher = self.publisher.clone();
+        let incoming_task: JoinHandle<Result<(), Box<dyn Error + Send + Sync>>> =
+            task::spawn(async move {
+                loop {
```

Comment:
this block is drifting out to the right quite a bit. Can it be split into a function?

Reply:
I tried to do that, but it's tricky with lifetimes. Maybe we can try refactoring this together?

Reply:
Can do.

I worry that if splitting this out is tricky then allowing the callback to publish arbitrary state changes might also be tricky. Are we using the Option<String> return value as a get out of jail free card that will make life hard for anyone who wants to do things for themselves?